crates/floresta-chain/src/pruned_utreexo/mod.rs

```diff
@@ -351,23 +355,119 @@ impl<T: BlockchainInterface> BlockchainInterface for Arc<T> {
     ) -> Result<Stump, Self::Error> {
         T::update_acc(self, acc, block, height, proof, del_hashes)
     }
-
     fn get_chain_tips(&self) -> Result<Vec<BlockHash>, Self::Error> {
         T::get_chain_tips(self)
     }
-
     fn validate_block(
         &self,
         block: &Block,
         proof: Proof,
-        inputs: HashMap<OutPoint, TxOut>,
+        inputs: UtxoMap,
         del_hashes: Vec<sha256::Hash>,
-        acc: Stump,
+        chain_time: &ChainTime,
     ) -> Result<(), Self::Error> {
-        T::validate_block(self, block, proof, inputs, del_hashes, acc)
+        T::validate_block(self, block, proof, inputs, del_hashes, chain_time)
     }
 
     fn get_fork_point(&self, block: BlockHash) -> Result<BlockHash, Self::Error> {
         T::get_fork_point(self, block)
     }
 }
+
+pub mod utxo_data {
+    use super::*;
+
+    pub struct UtxoData {
+        /// The transaction output that created this UTXO.
+        pub txout: bitcoin::TxOut,
+        //we need this mostly to validate relative-locktime.
+        /// The height of the block that the utxo was commited.
+        pub height: u64,
+        /// The timestamp of the block that the utxo was commited.
+        pub time: u32,
+    }
+
+    impl UtxoData {
+        /// Creates a new UtxoData.
+        pub const fn new(txout: bitcoin::TxOut, height: u64, time: u32) -> Self {
+            Self {
+                txout,
+                height,
+                time,
+            }
+        }
+        /// Gets the output.
+        pub const fn get_txout(&self) -> &bitcoin::TxOut {
+            &self.txout
+        }
+        /// Gets the height that the output was commited.
+        pub const fn get_height(&self) -> u64 {
+            self.height
+        }
+        pub const fn get_time(&self) -> u32 {
+            self.time
+        }
+    }
+    /// A [`HashMap<bitcoin::OutPoint, UtxoData>`] alias.
+    ///
+    /// Use HashMap to store and fast-find all needed UTXOs related data.
+    pub type UtxoMap = HashMap<bitcoin::OutPoint, UtxoData>;
+}
+
+pub mod chain_time {
+    /// We need a some local time notation while validating blocks
+    /// and a notation of the timestamp of the blocks.
+    ///
+    /// floresta-chain relies on the [`ChainTime`] to have some block validation.
+    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    pub struct ChainTime {
```

Comment:
Making a struct here will require you to create a new one every block you want to validate, looks like a lot of code duplication. Perhaps a trait would be a better option?

Reply:
> Perhaps a trait would be a better option?

Being generic over traits can carry us to undefined behavior and making a safe implementation has a lot of details to check on. 

I dont like that much the actual solution, but with a struct we can simplify things over the code. specially while testing.  I think its better than a trait, but thats on you to decide. 

If you prefer a trait, please submit an example or a point-to-start. One of the reasons why i chose build up a struct was because its a straight forward idea.

Reply:
We decided to use an external source of time because this crate should support `no-std`, and we can't get the wall time without the std-lib (modulo with some dependency, maybe). So I think this should laser-focused on returning to the current time. Something like this: 

```rust
/// An oracle that returns the current time for validation
pub trait ChainTime{
    fn current_time() -> u32;
}
```
and a concrete implementation using `std` simply uses system time, this can be built on `florestad` and passed to `FlorestaChain` on initialization. Every time we need time, we can just call this method. It offers the advantage of allowing for more sophisticated forms of getting consensus over the current time down the road because everything is abstracted away. `ChainState` only requires "something that gives me a reasonable account for the current wall time". 
Everything else (like MTP) comes from the chain itself, so I think it makes sense to leave it somewhere else. Furthermore, this time is likely only used once: to check whether the header time exceeds two hours into the future. All locktimes and the minimum time are computed using MTP.

This function may return an enum to indicate whether we want to validate time, just like you are doing here (which is a good idea).

Reply:
Perfect!